### PR TITLE
[X86][AMX] Support AMX-FP8

### PR DESCRIPTION
**Pull request by fzou1**
Ref.: https://cdrdv2.intel.com/v1/dl/getContent/671368

### CODE REVIEW

File: clang/lib/Headers/amxfp8intrin.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+/*===---------- amxfp8intrin.h - AMX intrinsics -*- C++ -*------------===
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:1]: Make it aligned to 80 column.

File: clang/lib/Headers/amxfp8intrin.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#endif /* __IMMINTRIN_H */
+
+#ifndef __AMXFP8INTRIN_H
+#define __AMXFP8INTRIN_H
+#ifdef __x86_64__
+
+#define _tile_dpbf8ps __builtin_ia32_tdpbf8ps
+#define _tile_dpbhf8ps __builtin_ia32_tdpbhf8ps
+#define _tile_dphbf8ps __builtin_ia32_tdphbf8ps
+#define _tile_dphf8ps __builtin_ia32_tdphf8ps
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:21]: List arguments for better reading.

File: llvm/lib/Target/X86/X86.td [hunk 3]
```diff
@@ -270,6 +270,9 @@ def FeatureAMXFP16     : SubtargetFeature<"amx-fp16", "HasAMXFP16", "true",
 def FeatureAMXCOMPLEX : SubtargetFeature<"amx-complex", "HasAMXCOMPLEX", "true",
                                          "Support AMX-COMPLEX instructions",
                                          [FeatureAMXTILE]>;
+def FeatureAMXFP8 : SubtargetFeature<"amx-fp8", "HasAMXFP8", "true",
+                                         "Support AMX-FP8 instructions",
+                                         [FeatureAMXTILE]>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86.td:275]: Align within `<`

File: llvm/lib/Target/X86/X86ISelLowering.cpp [hunk 4]
```diff
@@ -37503,6 +37503,38 @@ X86TargetLowering::EmitInstrWithCustomInserter(MachineInstr &MI,
     MI.eraseFromParent(); // The pseudo is gone now.
     return BB;
   }
+  case X86::PTDPBF8PS:
+  case X86::PTDPBHF8PS:
+  case X86::PTDPHBF8PS:
+  case X86::PTDPHF8PS: {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86ISelLowering.cpp:37509]: Combine with line 37413.

File: llvm/lib/Target/X86/X86InstrAMX.td [hunk 5]
```diff
@@ -267,3 +267,42 @@ let Predicates = [HasAMXCOMPLEX, In64BitMode] in {
...
+
+// AMX-FP8
+let Predicates = [HasAMXFP8, In64BitMode] in {
+  let SchedRW = [WriteSystem] in {
+    let Constraints = "$src1 = $dst" in {
+      class AMX_FP8_BASE<bits<8> Opcode, string Opstr> :
+        I<Opcode, MRMSrcReg4VOp3, (outs TILE:$dst),
+        (ins TILE:$src1, TILE:$src2, TILE:$src3),
+        !strconcat(Opstr, "\t{$src3, $src2, $dst|$dst, $src2, $src3}"),
+        []>, VEX, VVVV;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAMX.td:279]: Align within `<`

File: llvm/lib/Target/X86/X86InstrAMX.td [hunk 6]
```diff
@@ -267,3 +267,42 @@ let Predicates = [HasAMXCOMPLEX, In64BitMode] in {
...
+                                        [(int_x86_tdpbhf8ps timm:$src1,
+                                          timm:$src2, timm:$src3)]>;
+      def PTDPHBF8PS : PseudoI<(outs), (ins u8imm:$src1,
+                                        u8imm:$src2, u8imm:$src3),
+                                        [(int_x86_tdphbf8ps timm:$src1,
+                                          timm:$src2, timm:$src3)]>;
+      def PTDPHF8PS : PseudoI<(outs), (ins u8imm:$src1,
+                                       u8imm:$src2, u8imm:$src3),
+                                       [(int_x86_tdphf8ps timm:$src1,
+                                         timm:$src2, timm:$src3)]>;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86InstrAMX.td:305]: ditto.

File: llvm/lib/TargetParser/Host.cpp [hunk 7]
```diff
@@ -1876,6 +1876,10 @@ const StringMap<bool> sys::getHostCPUFeatures() {
       MaxLevel >= 0x19 && !getX86CpuIDAndInfo(0x19, &EAX, &EBX, &ECX, &EDX);
   Features["widekl"] = HasLeaf7 && HasLeaf19 && ((EBX >> 2) & 1);
 
+  bool HasLeaf1E =
+      MaxLevel >= 0x1e && !getX86CpuIDAndInfo(0x1e, &EAX, &EBX, &ECX, &EDX);
```
Comment (phoebewang) [inline llvm/lib/TargetParser/Host.cpp:1880]:
```
  bool HasLeaf1E = MaxLevel >= 0x1e &&
                   !getX86CpuIDAndInfoEx(0x1e, 0x1, &EAX, &EBX, &ECX, &EDX);
```

File: llvm/test/MC/Disassembler/X86/AMX/x86-64-amx-fp8.txt [hunk 8]
```diff
@@ -0,0 +1,34 @@
+# RUN: llvm-mc --disassemble %s -triple=x86_64 | FileCheck %s --check-prefixes=ATT
```
Comment (phoebewang) [inline llvm/test/MC/Disassembler/X86/AMX/x86-64-amx-fp8.txt:1]: Remove `x86-64-` from file name. The same below.

File: clang/lib/Headers/amxfp8intrin.h [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+///
+/// This intrinsic corresponds to the \c TDPBF8PS instruction.
+///
+/// \param dst
+///    The destination tile. Max size is 1024 Bytes.
+/// \param a
+///    The 1st source tile. Max size is 1024 Bytes.
+/// \param b
+///    The 2nd source tile. Max size is 1024 Bytes.
+#define _tile_dpbf8ps __builtin_ia32_tdpbf8ps
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:38]: `#define _tile_dpbf8ps(dst, a, b) __builtin_ia32_tdpbf8ps((dst), (a), (b))`
The same below.

File: clang/lib/Headers/amxfp8intrin.h [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+#ifndef __AMXFP8INTRIN_H
+#define __AMXFP8INTRIN_H
+#ifdef __x86_64__
+
+
+/// Compute dot-product of brain-float8 (BF8) or hybrid-float8 (HF8)
+///    floating-point pairs in tiles \a a and \a b, accumulating the
+///    intermediate single-precision (32-bit) floating-point elements with
+///    elements in \a dst, and store the 32-bit result back to tile \a dst.
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:22]: The comment need to be added for all intrinsics.

Comment (fzou1) [reply clang/lib/Headers/amxfp8intrin.h]: Okay. I'll copy description from instruction set extension document for the file and each intrinsic.

File: llvm/include/llvm/IR/IntrinsicsX86.td [hunk 11]
```diff
@@ -5994,6 +5994,23 @@ let TargetPrefix = "x86" in {
...
+                        [ImmArg<ArgIndex<0>>,
+                         ImmArg<ArgIndex<1>>, ImmArg<ArgIndex<2>>]>;
+  def int_x86_tdphbf8ps : ClangBuiltin<"__builtin_ia32_tdphbf8ps">,
+              Intrinsic<[], [llvm_i8_ty, llvm_i8_ty, llvm_i8_ty],
+                        [ImmArg<ArgIndex<0>>,
+                         ImmArg<ArgIndex<1>>, ImmArg<ArgIndex<2>>]>;
+  def int_x86_tdphf8ps : ClangBuiltin<"__builtin_ia32_tdphf8ps">,
+              Intrinsic<[], [llvm_i8_ty, llvm_i8_ty, llvm_i8_ty],
+                        [ImmArg<ArgIndex<0>>,
+                        ImmArg<ArgIndex<1>>, ImmArg<ArgIndex<2>>]>;
```
Comment (phoebewang) [inline llvm/include/llvm/IR/IntrinsicsX86.td:6013]: Move them to line 5919. This is for internal intrinsics.

File: clang/lib/Headers/amxfp8intrin.h [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+#ifndef __AMXFP8INTRIN_H
+#define __AMXFP8INTRIN_H
+#ifdef __x86_64__
+
+/// These instructions compute dot product of brain-float8 (BF8) or
+/// hybrid-float8 (HF8) accumulating into a single precision (FP32). The input
+/// elements can be BF8 or HF8. These instructions have three tile operands, one
+/// source/dest accumulator operand, and two source operands, \a a and \a b. The
+/// \a a and \a b operands can be BF8 or HF8 independently, and the source/dest
+/// operand, \a dst is always FP32.
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:23]: We usually don't add description for the whole file. I think we can simply remove it, ot If you want, you can follow the way like https://github.com/llvm/llvm-project/blob/main/clang/lib/Headers/stddef.h#L10-L21

File: clang/lib/Headers/amxfp8intrin.h [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+///
+/// \headerfile <immintrin.h>
+///
+/// \code
+/// void _tile_dpbf8ps (__tile dst, __tile a, __tile b)
+/// \endcode
+///
+/// This intrinsic corresponds to the \c TDPBF8PS instruction, which is the dot
+/// product of a BF8 value (\a a) by a BF8 value (\a b) accumulating into a
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:32]: Do not use parentheses around `\a a`

File: clang/lib/Headers/amxfp8intrin.h [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+/// This intrinsic corresponds to the \c TDPBHF8PS instruction, which is the dot
+/// product of a BF8 value (\a a) by an HF8 value (\a b) accumulating into a
+/// Single Precision (FP32) source/dest (\a dst).
+///
+/// \param dst
+///    The destination tile. Max size is 1024 Bytes.
+/// \param a
+///    The 1st source tile. Max size is 1024 Bytes.
+/// \param b
+///    The 2nd source tile. Max size is 1024 Bytes.
```
Comment (phoebewang) [inline clang/lib/Headers/amxfp8intrin.h:56]: The current rule of rule of thumb is
```
Intrinsic description

Pseudo code if any

/// \headerfile <immintrin.h>

/// \code
/// void _tile_dpbhf8ps (__tile dst, __tile a, __tile b)
/// \endcode

/// This intrinsic corresponds to the \c TDPBHF8PS instruction.

/// \param dst
///    The destination tile. Max size is 1024 Bytes.
/// \param a
///    The 1st source tile. Max size is 1024 Bytes.
/// \param b
///    The 2nd source tile. Max size is 1024 Bytes.
```
See examples in other files.

File: clang/lib/Headers/amxfp8intrin.h (hunk 12 above)
Comment (fzou1) [reply clang/lib/Headers/amxfp8intrin.h]: Okay. I'll remove it.

File: clang/lib/Headers/amxfp8intrin.h (hunk 14 above)
Comment (fzou1) [reply clang/lib/Headers/amxfp8intrin.h]: Thanks. Updated.

**Review by phoebewang — APPROVED** [review]
LGTM, need to add new api support as a follow up.

**Comment by fzou1** [comment]
> LGTM, need to add new api support as a follow up.

Got it. I'll add it later. Please help merge. Thanks.

**Pull request merged**